### ISSUE TITLE
Configure Does Not Seem to Work Correctly

### ISSUE THREAD

**Issue by nmhamster** [trilinos/Trilinos#2324]
@bartlettroscoe - getting the weirdness below. It looks like the directory check isn't correct?

```
$ ../../do-configure-devpack-intel-18.1.163.sh
rm: cannot remove ‘CMakeCache.txt’: No such file or directory

Configuring Trilinos build directory

CMake Error at cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake:88 (MESSAGE):
  ERROR!
  CMAKE_CURRENT_SOURCE_DIR=/ascldap/users/sdhammo/git/trilinos-github-repo ==
  CMAKE_CURRENT_BINARY_DIR=/ascldap/users/sdhammo/git/trilinos-github-repo

  Trilinos does not support in source builds!

  NOTE: You must now delete the CMakeCache.txt file and the CMakeFiles/
  directory under the source directory for Trilinos or you will not be able
  to configure Trilinos correctly!

  You must now run something like:

    $ rm -r CMakeCache.txt CMakeFiles/

  Please create a different directory and configure Trilinos under that such
  as:

    $ mkdir MY_BUILD
    $ cd MY_BUILD
    $ cmake [OPTIONS] ..
Call Stack (most recent call first):
  cmake/tribits/core/package_arch/TribitsProjectImpl.cmake:91 (TRIBITS_ASSERT_AND_SETUP_PROJECT_AND_STATIC_SYSTEM_VARS)
  cmake/tribits/core/package_arch/TribitsProject.cmake:93 (TRIBITS_PROJECT_IMPL)
  CMakeLists.txt:93 (TRIBITS_PROJECT)


-- Configuring incomplete, errors occurred!
See also "/ascldap/users/sdhammo/git/trilinos-github-repo/CMakeFiles/CMakeOutput.log".
-bash-4.2$ pwd
/ascldap/users/sdhammo/git/trilinos-github-repo/build-devpack-intel-181163-mini
```

**Comment by bartlettroscoe** [comment]
@nmhamster, are you doing an in-source build?  That is what it looks like from:

```
ERROR!
  CMAKE_CURRENT_SOURCE_DIR=/ascldap/users/sdhammo/git/trilinos-github-repo ==
  CMAKE_CURRENT_BINARY_DIR=/ascldap/users/sdhammo/git/trilinos-github-repo
```

To fix this, just use a different directory for you binary directory as the error message suggests.

And note that your script needs to remove the base `CMakeFiles/` directory in addition to the `CMakeCache.txt` file.  See:

*  https://tribits.org/doc/TribitsBuildReference.html#reconfiguring-completely-from-scratch

**Comment by nmhamster** [comment]
No, look at the directory at the bottom (the `pwd`). I have a directory inside the source where I build, so I make a directory, go into and try to start cmake just like I always have done. This is super annoying.

**Comment by bartlettroscoe** [comment]
What happens when you type:

```
$ rm -r CMake*
$ cmake ..
```

?

**Comment by nmhamster** [comment]
@bartlettroscoe - ping

**Comment by nmhamster** [comment]
@bartlettroscoe - same behavior as before. Has the code which performs this check changed? Everything was working fine. Very odd.

**Comment by bartlettroscoe** [comment]
> same behavior as before. Has the code which performs this check changed? Everything was working fine. Very odd.

Strange.  I just tried that and it worked fine.  We may need to pair program this together.

**Comment by nmhamster** [comment]
@bartlettroscoe - this has started happening on all sorts of machines now. Its effectively broken our ability to get any work done. Do you have any ideas at all?

```
$ ../../do-configure-devpack-intel-18.1.163.sh
rm: cannot remove ‘CMakeCache.txt’: No such file or directory

Configuring Trilinos build directory

CMake Error at cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake:88 (MESSAGE):
  ERROR!
  CMAKE_CURRENT_SOURCE_DIR=/ascldap/users/sdhammo/git/trilinos-github-repo ==
  CMAKE_CURRENT_BINARY_DIR=/ascldap/users/sdhammo/git/trilinos-github-repo

  Trilinos does not support in source builds!

  NOTE: You must now delete the CMakeCache.txt file and the CMakeFiles/
  directory under the source directory for Trilinos or you will not be able
  to configure Trilinos correctly!

  You must now run something like:

    $ rm -r CMakeCache.txt CMakeFiles/

  Please create a different directory and configure Trilinos under that such
  as:

    $ mkdir MY_BUILD
    $ cd MY_BUILD
    $ cmake [OPTIONS] ..
Call Stack (most recent call first):
  cmake/tribits/core/package_arch/TribitsProjectImpl.cmake:91 (TRIBITS_ASSERT_AND_SETUP_PROJECT_AND_STATIC_SYSTEM_VARS)
  cmake/tribits/core/package_arch/TribitsProject.cmake:93 (TRIBITS_PROJECT_IMPL)
  CMakeLists.txt:93 (TRIBITS_PROJECT)


-- Configuring incomplete, errors occurred!
See also "/ascldap/users/sdhammo/git/trilinos-github-repo/CMakeFiles/CMakeOutput.log".
```

But ..

```
$ pwd
/ascldap/users/sdhammo/git/trilinos-github-repo/build-devpack-intel-181163
```

Note - I am in a clean sub-directory. I've blown all the files away and started clean. It looks like the directory detection is screwed up.

**Comment by nmhamster** [comment]
@bartlettroscoe - worked it out. There was a `CMakeCache.txt` in the path in the Trilinos directory that was screwing it up (essentially in Trilinos source). I'll close the issue.

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe - worked it out. There was a CMakeCache.txt in the path in the Trilinos directory that was screwing it up (essentially in Trilinos source). I'll close the issue.

That is good to know.  That is one example of why you don't want to be building in the source tree.  For me, I always have my build dirs in parallel from the Trilinos source tree like:

```
<some-base-dir>/
    Trilinos/
    BUILDS/
        GCC-4.8.4/
           MPI_DEBUG/
           ...
        ...
```

Just a suggestion.

**Comment by mhoemmen** [comment]
I was going to open an issue for detecting this, but Ross' error message explains exactly what's going on and how to fix it ;-) .

**Comment by bartlettroscoe** [comment]
> I was going to open an issue for detecting this, but Ross' error message explains exactly what's going on and how to fix it ;-) .

But it sounds like after that initial `CMakeCache.txt` file is created that even when you move into the new directory it will still generate that error message, which is misleading at that point.

Perhaps we need to update TriBITS to look for the file `${PROJECT_SORUCE_DIR}/CMakeCache.txt` and if it exists, error out right away with a specialized error message like:

```
$ ../../do-configure-devpack-intel-18.1.163.sh
rm: cannot remove ‘CMakeCache.txt’: No such file or directory

Configuring Trilinos build directory

CMake Error at cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake:??? (MESSAGE):
  ERROR!
  the file /ascldap/users/sdhammo/git/trilinos-github-repo/CMakeCache.txt exists in the base
  source tree!

  Trilinos does not support in source builds!

  NOTE: You must now delete the CMakeCache.txt file and the CMakeFiles/
  directory under the source directory for Trilinos or you will not be able
  to configure Trilinos correctly!

  You must now run something like:

    $ cd /ascldap/users/sdhammo/git/trilinos-github-repo/
    $ rm -r CMakeCache.txt CMakeFiles/

  Please create a different directory and configure Trilinos under that such
  as:

    $ mkdir MY_BUILD
    $ cd MY_BUILD
    $ cmake [OPTIONS] ..
```

Would that help?  That would not be too much work to add to TriBITS.

**Comment by nmhamster** [comment]
@bartlettroscoe - do you think it might be possible to say where it found the `CMakeCache.txt` so the user can identify what `cmake` actually did?

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe - do you think it might be possible to say where it found the CMakeCache.txt so the user can identify what cmake actually did?

@nmhamster,

Yup.  That is what I showed in the prototype output [above](https://github.com/trilinos/Trilinos/issues/2324#issuecomment-370491862) which shows:

```
CMake Error at cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake:??? (MESSAGE):
  ERROR!
  the file /ascldap/users/sdhammo/git/trilinos-github-repo/CMakeCache.txt exists in the base
  source tree!
```

Does that make sense?

**Comment by nmhamster** [comment]
@bartlettroscoe - this looks great, it would be helpful.

**Comment by mhoemmen** [comment]
Thanks Ross! :D

**Comment by bartlettroscoe** [comment]
PR #2341 
 adds a better error message for this situation.  Here is what it looks like for Trilinos:

```
[rabartl@crf450 BUILD (develop)]$ cmake ..

Configuring Trilinos build directory

CMake Error at cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake:101 (MESSAGE):
  ERROR! The file
  /ascldap/users/rabartl/Trilinos.base/Trilinos/CMakeCache.txt exists from a
  likely prior attempt to do an in-soruce build.

  You must now run something like:

    $ cd /ascldap/users/rabartl/Trilinos.base/Trilinos/
    $ rm -r CMakeCache.txt CMakeFiles/

  Please create a different directory and configure Trilinos under that such
  as:

    $ cd /ascldap/users/rabartl/Trilinos.base/Trilinos/
    $ mkdir MY_BUILD
    $ cd MY_BUILD
    $ cmake [OPTIONS] ..
Call Stack (most recent call first):
  cmake/tribits/core/package_arch/TribitsProjectImpl.cmake:91 (TRIBITS_ASSERT_AND_SETUP_PROJECT_AND_STATIC_SYSTEM_VARS)
  cmake/tribits/core/package_arch/TribitsProject.cmake:93 (TRIBITS_PROJECT_IMPL)
  CMakeLists.txt:93 (TRIBITS_PROJECT)


-- Configuring incomplete, errors occurred!
```

@nmhamster and @mhoemmen, does that look okay?

The auto PR testing started over 2 hours ago but is still running.  Once it complete, I will push.

**Comment by bartlettroscoe** [comment]
I went ahead and merged #2341 since the tests passed and this is a small set of changes.

**Comment by mhoemmen** [comment]
@bartlettroscoe Looks good, just fix spelling:

> exists from a likely prior attempt to do an in-soruce build.

thanks Ross! :D

**Comment by bartlettroscoe** [comment]
@mhoemmen, typo is resolved in tribitspub/tribits@b8a93336514c9b825ace9d286eb18cda8d96e1e5.  It will get updated in Trilinos the next time TriBITS get snapshotted (likely tomorrow).

Can we close this issue now?

**Comment by mhoemmen** [comment]
@bartlettroscoe Thanks Ross! You're awesome! :D:D:D
